addons/mail/models/mail_thread.py

```diff
@@ -835,6 +836,75 @@ def _routing_reset_bounce(self, email_message, message_dict):
             for model in [bl_model for bl_model in bl_models if bl_model.model in self.env]:  # transient test mode
                 self.env[model.model].sudo().search([('message_bounce', '>', 0), ('email_normalized', '=', valid_email)])._message_reset_bounce(valid_email)
 
+    @api.model
+    def _routing_detect_loop_model(self, message_dict, routes):
```

Comment:
After checking new code, new suggestion ! We can call this method ``_routing_detect_loop_from_records`` (to distinguish from the other one that is "from_headers"). Then this one could call a model-based sub method for each route (``def _routing_detect_duplicates(self, email)`` which would be a true multi ?), that we could override in ``mail.thread.blacklist`` and use email_normalized. That way it is cleanly modular (don't have to "guess" the inheritance), and allow custom behavior based on model (people may override this method to have their own duplicate check if necessary).

It gives a bit more methods (3 in the end) but based on previous mail.thread iterations, it is the best way to offer flexibility to sub models while being modular. What do you think ?